### DISCUSSION TITLE
Code Template:  Execute cURL Post in Transformer

### DISCUSSION THREAD

**Discussion by rjkroll** [nextgenhealthcare/connect#5295] [General]
Hi,

I thought maybe someone would find this useful.   We had an HTTP Sender that had been working fine for years.   The receiving end made an update that broke our file POST submissions, were taking forever to look into it, and we knew cURL submissions worked, so we just built a quick little function to send a POST through cURL command line,  through a transformer.    Not entirely sure why cURL works and the HTTP Sender suddenly didnt, but here is the function:

Source:  Normal File reader, normal Move-to-directly/Move-to-Error Directory
Destination:  Javascript Writer with the code below

What it does: 

- Takes the files from the source directly and POSTS them via cURL to an endpoint
- Gets the response (in this case JSON) and sends it to a different channel for processing
-  Forces either a true Mirth SENT/ERROR status, if ERROR the file moves to "Error Directory Folder" designated in Source
-  You can modify the 'cmd' line to add any necessary cURL flags you need for your interface
-  Function assumes you have GIT installed, otherwise point to local cURL.exe of your machine
-  This function is built for using JWT token for Authorization, but can be modified for any Auth type

```
function runCurl(fileName,filePath,token){
	var path = filePath;
	if(!filePath.endsWith("\\")){ path +="\\";}
	var cmd = 'C:\\Program Files\\Git\\mingw64\\bin\\curl.exe curl -X POST -H "x-file-name:'+ fileName + 
	'" -H "Authorization: ' + token + '" -F "file=@' +path + fileName + '" https://<endpoint url here>';
	logger.info("command:" + cmd);
	var result = executeRuntimeCommand(cmd);	
	logger.info('Exit value: ' + result.exitValue);
	logger.info('Output: ' + result.stdout);
	if (result.stderr != '') {
		logger.error('Errors: ' + result.stderr);
	}
	channelMap.put('curlOutput',result.stdout);

	channelMap.put('curlExit',result.exitValue);	
	channelMap.put('curlStdErr',result.stderr);	
	if(parseInt(result.exitValue,10) !== 0){return false;/*"Error in curl command"*/}		
	router.routeMessageByChannelId('166476d0-aacf-4e06-9553-2da83d900b37', result.stdout);
	return true;
}

var curlResult = runCurl(sourceMap.get('originalFilename'),sourceMap.get('fileDirectory'), globalMap.get('JWTtoken'));
if(!curlResult){return ERROR;}
return SENT;
```

**Comment by pacmano1** [comment]
Interesting.   Why not send the curl post that works and the mirth one that does not work to a dumb http echo server and  compare the differences?  

This echo server is handy: https://www.npmjs.com/package/http-echo-server

**Reply by rjkroll** [reply]
We actually did do that because I saw a post of yours before were you mentioned spinning up a local server for the same reason.   We spent hours trying to modify the HTTP Sender, even wrote out the full cURL --trace and modified our headers/content...everything to be perfectly the same, but still couldn't POST correctly.   Their server was CLEARLY returning a Python error when receiving our file and they wouldn't look into it for days, so we needed to come up with another solution.     As to why cURL worked and Postman worked, but the interface in Mirth that had been setup for 2 years stopped working, still waiting to hear back from them about it.

**Reply by pacmano1** [reply]
Interesting.  Helpful post of course.
